### ISSUE TITLE
Disable level selector after start

### ISSUE THREAD

**Issue by Blaisorblade** [farkam135/GoIV#467]
A friend just raised their level after GoIV was started—and of course the level wasn't really updated.

He suggested disabling the level changer at that point as an alternative to updating the level for real. I'd propose to do this fix quickly, before we try the more complicated "reflect new trainer level throughout the code" (that's also possible but trickier and more bug-prone).

**Comment by NightMadness** [comment]
Options
1. Disable scroll selector when GoIV "Start" button is pressed
2. Stop and start on trainer level change completed.
3. Recalculate only whats needed on trainer level change completed, (not restart entire GoIV).

I think we should go with option 2 stop/start on level change and see if it works properly if slider doesnt have a good on "complete event" we can go with option 1 or be fancy and have a mini timer.

On a slde note, I think the scoll selector is useful the first time we setup GoIV but we really should have + - buttons since normal usage is only move by one.

**Comment by Blaisorblade** [comment]
I'm specifically proposing option 1 as a quickfix. Restarting GoIV is IMHO too visibly slow for option 2 to feel good to use (even on my Nexus 5X, and there are slower phones).

We can skip 1 if somebody manages to implement option 3 correctly before we do 3, but it's not trivial.

> On a side note, I think the scoll selector is useful the first time we setup GoIV but we really should have + - buttons since normal usage is only move by one.

I'd tend to agree but the scroll selector seems easy to increment/decrement. (I'd love a +1 button in the notification though it might be overkill).

**Comment by sarav** [comment]
I'm okay with 1 as as quick fix. Makes sense. A toast to explain might be better. But I think 2 is also a good quick fix. By disabling trainer level change, you are effectively having the user stop, change level and then start the service again. So, might as well just do it for them till we can figure out a way to just recalculate stuff.

**Comment by sarav** [comment]
I actually never liked the scroll selector. I'd rather just have an input box with +/- like we had before. Or a scroll bar like we had for the level selector in the results window. As for notification @thearaks made a good point that we shouldn't clutter the notif with stuff that won't be used often (I don't have a strong opinion on this).

**Comment by Blaisorblade** [comment]
> By disabling trainer level change, you are effectively having the user stop, change level and then start the service again. So, might as well just do it for them till we can figure out a way to just recalculate stuff.

It's not the same, unless you only do the restart when they're actually done selecting the level. It takes 2 seconds for a wheel to be done scrolling?

I see there are downsides the other way.

> I actually never liked the scroll selector. I'd rather just have an input box with +/- like we had before.

We didn't have +/- there IIRC—we requested that and that's what @oskomonzon did. I also preferred normal +/-, not sure why he had this preference.

**Comment by DJCrapsody** [comment]
The current number scroll selector is actually Android default number picker (as of ICS, iirc). Maybe we can change to a simple one with +/-.

**Comment by nahojjjen** [comment]
Is this issue a newer version of #241 ?

**Comment by sarav** [comment]
@nahojjjen, It's evolving (or devolving?) into that one. 

@Blaisorblade -- Yes, I meant once the user is done selecting it. We can always do the poor mans job of doing this start stop a few seconds after they are done changing it? So, if they continue scrolling, we won't take 2 seconds for each increment. We'll have to figure out the delay or something like that even if we implement (3) I think.

**Comment by NightMadness** [comment]
I think https://github.com/farkam135/GoIV/issues/241 need to be reopened, it's about reloading settings like "screenshots to calculate iv" setting or any other settings, this bug is about disabling selector.

I'm playing around with the setOnScrollListener looks like it will work but if you scroll all the way from 1 you will stop scrolling at some point for just a moment so a timer may be necessary. setOnScrollListener will not work for changing text with the keyboard. We don't have to but we should (but don't have to??) wait for stop to be done before starting.

Anyway, this is the code to enable/disable level selector on start:
 ```
   /**
     * Starts the PokeFly background service which contains overlay logic.
     */
    private void startPokeFly() {
        launchButton.setText(R.string.main_starting);
        launchButton.setEnabled(false);

        int statusBarHeight = getStatusBarHeight();
        Intent intent = Pokefly.createIntent(this, trainerLevel, statusBarHeight, batterySaver);
        startService(intent);

        if (settings.shouldLaunchPokemonGo()) {
            openPokemonGoApp();
        }
        enableDisableTrainerLevelPicker(false);
    }

    private void updateLaunchButtonText(boolean isPokeflyRunning) {
        if (!hasAllPermissions()) {
            launchButton.setText(R.string.main_permission);
        } else if (isPokeflyRunning) {
            launchButton.setText(R.string.main_stop);
        } else {
            launchButton.setText(R.string.main_start);
        }
        enableDisableTrainerLevelPicker(!isPokeflyRunning);
    }

    /**
     * enables/disabled trainer level picker.
     */
    private void enableDisableTrainerLevelPicker(boolean enabled) {
        final NumberPicker npTrainerLevel = (NumberPicker) findViewById(R.id.trainerLevel);
        npTrainerLevel.setEnabled(enabled);
    }
```

------
P.S. code for setOnScrollListener
--> if we want to use this it Needs work
1. Doesn't work for keyboard edit.
2. Should extract start/stop function from button instead of copying.
3. should wait for stop to be completed?
4. add timer (how long?)
5. disable picker during "Starting..."
```
            npTrainerLevel.setOnScrollListener(new NumberPicker.OnScrollListener() {
            @Override
            public void onScrollStateChange(NumberPicker numberPicker, int scrollState) {
                if (scrollState == SCROLL_STATE_IDLE && !numberPicker.hasFocus() && Pokefly.isRunning()) {
                    stopService(new Intent(MainActivity.this, Pokefly.class));
                    if (screen != null) {
                        screen.exit();
                    }
                    batterySaver = settings.isManualScreenshotModeEnabled();
                    setupDisplaySizeInfo();
                    trainerLevel = setupTrainerLevel(npTrainerLevel);

                    Data.setupArcPoints(arcInit, arcRadius, trainerLevel);

                    if (batterySaver) {
                        startPokeFly();
                    } else {
                        startScreenService();
                    }

                }
            }
        });
```

**Comment by sarav** [comment]
@NightMadness

* Just call it enableTrainerLevelPicker(boolean enable) or toggleTrainerLevelPicker(). enableDisable... is way too wordy.
* You are enabling the picker anytime you update the button text. It should not be that way. You should only enable it only when it goes back to Start.

**Comment by sarav** [comment]
Also, the timer and waiting for stop before starting (we absolutely have to) can all be handled by just doing a postMessage and postMessageDelayed. They all go through the same queue -- so they'll automatically be ordered and serialized.

**Comment by Blaisorblade** [comment]
FWIW, if I finish and we merge https://github.com/farkam135/GoIV/tree/refactor-screen-info, which encapsulates the screen- and trainer-level specific info, propagating a new trainer level might get easier (not 100% trivial).
